File: src/commands/install_lazy.cpp
#include "base.hpp"
#include "commands.hpp"
#include "lazy_config.hpp"

#include <filesystem>
#include <fstream>
#include <iostream>

namespace commands {
void install_lazy() {
  std::cout << "Lazy installation for a lazy plugin manager...\n";

  std::string lazy_file_path = base::nvPath.config_path + "/lazy.lua";

  try {
    if (!std::filesystem::exists(base::nvPath.lua_path))
      std::filesystem::create_directory(base::nvPath.lua_path);

    if (!std::filesystem::exists(base::nvPath.config_path))
      std::filesystem::create_directory(base::nvPath.config_path);

    if (!std::filesystem::exists(lazy_file_path)) {
      std::ofstream file(lazy_file_path);
      if (!file.is_open())
        throw std::runtime_error("Failed to create lazy.lua");
      file << LAZY_LUA_CONTENT;
      file.close();
    }

    if (!std::filesystem::exists(base::nvPath.plugins_path))
      std::filesystem::create_directory(base::nvPath.plugins_path);

    std::ofstream init(base::nvPath.init_path, std::ios::app);
    if (!init.is_open())
      throw std::runtime_error("Failed to open init.lua");
    init << "\nrequire(\"config.lazy\")\n";
    init.close();

    std::cout << "Lazy installed successfully!\n";

  } catch (const std::exception &e) {
    std::cerr << "Error: " << e.what() << " — rolling back...\n";

    // rollback
    if (std::filesystem::exists(lazy_file_path))
      std::filesystem::remove(lazy_file_path);

    if (std::filesystem::exists(base::nvPath.plugins_path))
      std::filesystem::remove(base::nvPath.plugins_path);

    if (std::filesystem::exists(base::nvPath.config_path))
      std::filesystem::remove(base::nvPath.config_path);

    if (std::filesystem::exists(base::nvPath.lua_path))
      std::filesystem::remove(base::nvPath.lua_path);

    std::cerr << "Rollback complete.\n";
  }
}
} // namespace commands

File: go/commands/install_lazy.go
package commands

import (
	"lpm/base"
	"os"
)

func lazyBootstrap() string {
	return `-- Bootstrap lazy.nvim
local lazypath = vim.fn.stdpath("data") .. "/lazy/lazy.nvim"
if not (vim.uv or vim.loop).fs_stat(lazypath) then
  local lazyrepo = "https://github.com/folke/lazy.nvim.git"
  local out = vim.fn.system({ "git", "clone", "--filter=blob:none", "--branch=stable", lazyrepo, lazypath })
  if vim.v.shell_error ~= 0 then
    vim.api.nvim_echo({
      { "Failed to clone lazy.nvim:\n", "ErrorMsg" },
      { out, "WarningMsg" },
      { "\nPress any key to exit..." },
    }, true, {})
    vim.fn.getchar()
    os.exit(1)
  end
end
vim.opt.rtp:prepend(lazypath)

vim.g.mapleader = " "
vim.g.maplocalleader = "\\"

require("lazy").setup({
  spec = {
    { import = "plugins" },
  },
  install = { colorscheme = { "habamax" } },
  checker = { enabled = true },
})`
}

func InstallLazy() {

	base.LpmLog("lazy installation for lazy plugin manager...", base.Ok)
	if !base.DirExists(base.NvPath.LuaPath + "/config") {
		err := os.Mkdir(base.NvPath.LuaPath+"/config", 0755)

		if err != nil {
			base.LpmLog(err.Error(), base.Err)
			return
		}
	}

	lazyFd, lazyFileErr := os.Create(base.NvPath.LuaPath + "/config/lazy.lua")
	if lazyFileErr != nil {
		base.LpmLog(lazyFileErr.Error(), base.Err)
		return
	}

	lazyFd.WriteString(lazyBootstrap())
	base.LpmLog("done.", base.Ok)
}
